### PR TITLE
fix: [WIN-NPM] process updatePods in fifo order

### PR DESCRIPTION
**Pull request by huntergregory**
**Reason for Change**:
Fully solve #1729. When the issue’s scenario occurs, ordering of Pods matters for UpdatePod() within ApplyDataPlane().

UTs have been failing for sequence 1 and 2 of this issue, and the chance of failure is higher when applying IPSets in background.

**Issue Fixed**:

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [x] adds unit tests

**Notes**:
### Solution
Handle `updatePod` objects in FIFO order, the same order as the control plane.

### Results
Update Pod ACLs in same order as the control plane, e.g., the first Pod created is the first Pod to have proper connectivity.

### CODE REVIEW

**Comment by huntergregory** [comment]
Closing because this issue would be extremely rare. All of these rare events would have to coincide.
1. Sequence 2 of #1729
2. Pods in question are on the same node
3. RefreshEndpoints failure

**Comment by huntergregory** [comment]
> Closing because this issue would be extremely rare. All of these rare events would have to coincide.
> 
> 1. Sequence 2 of [Edge Case for memory-starved Windows Server 2022 Nodes with NPM #1729](https://github.com/Azure/azure-container-networking/issues/1729)
> 2. Pods in question are on the same node
> 3. RefreshEndpoints failure

The chance of this occurring is higher when applying ipsets in background (bullet point 3 is no longer required).

Have now seen sequence 1 fail too.

File: npm/pkg/dataplane/dataplane.go [hunk 1]
```diff
@@ -149,6 +174,9 @@ func (dp *DataPlane) AddToSets(setNames []*ipsets.IPSetMetadata, podMetadata *Po
 			klog.Infof("[DataPlane] {AddToSet} pod key %s not found in updatePodCache. creating a new obj", podMetadata.PodKey)
 			updatePod = newUpdateNPMPod(podMetadata)
 			dp.updatePodCache.cache[podMetadata.PodKey] = updatePod
+
+			// add to queue only if not in the cache/queue already
+			dp.updatePodCache.queue = append(dp.updatePodCache.queue, podMetadata.PodKey)
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane.go:179]: Where are we checking if the key is not in queue ? we will need to have a 2nd check here,

File: npm/pkg/dataplane/dataplane.go [hunk 2]
```diff
@@ -177,6 +205,9 @@ func (dp *DataPlane) RemoveFromSets(setNames []*ipsets.IPSetMetadata, podMetadat
 			klog.Infof("[DataPlane] {RemoveFromSet} pod key %s not found in updatePodCache. creating a new obj", podMetadata.PodKey)
 			updatePod = newUpdateNPMPod(podMetadata)
 			dp.updatePodCache.cache[podMetadata.PodKey] = updatePod
+
+			// add to queue only if not in the cache/queue already
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane.go:209]: need to add a secondary check here if the key is in queue or not,

File: npm/pkg/dataplane/dataplane.go [hunk 3]
```diff
@@ -236,9 +268,18 @@ func (dp *DataPlane) ApplyDataPlane() error {
 		// lock updatePodCache while driving goal state to kernel
 		// prevents another ApplyDataplane call from updating the same pods
 		dp.updatePodCache.Lock()
-		defer dp.updatePodCache.Unlock()
+		defer func() {
+			dp.updatePodCache.removeDeletedItemsFromQueue()
+			dp.updatePodCache.Unlock()
```
Comment (vakalapa) [inline npm/pkg/dataplane/dataplane.go:273]: May be create a datastructure called ordered Map ? and use that as the update pod cache ? 

example: https://github.com/iancoleman/orderedmap/blob/master/orderedmap.go

File: npm/pkg/dataplane/types.go [hunk 4]
```diff
@@ -76,3 +78,70 @@ func (npmPod *updateNPMPod) updateIPSetsToRemove(setNames []*ipsets.IPSetMetadat
...
+		cache:           make(map[string]*updateNPMPod),
+		queue:           make([]string, 0, initialCapacity),
+		initialCapacity: initialCapacity,
+	}
+}
+
+// enqueue adds a pod to the queue if necessary and returns the pod object used
+func (c *updatePodCache) enqueue(m *PodMetadata) *updateNPMPod {
+	pod, ok := c.cache[m.PodKey]
+	if !ok {
```
Comment (vakalapa) [inline npm/pkg/dataplane/types.go:105]: What if the nodename got updated ? we will need to check if the pod objs are same if not update the element in map and move the pod obj to the end ?

File: npm/pkg/dataplane/types.go [hunk 5]
```diff
@@ -76,3 +78,70 @@ func (npmPod *updateNPMPod) updateIPSetsToRemove(setNames []*ipsets.IPSetMetadat
...
+	}
+
+	return pod
+}
+
+// dequeue returns the first pod in the queue and removes it from the queue.
+// Caller should ensure the queue is not empty.
+// Otherwise, the following will occur: "panic: runtime error: index out of range [0] with length 0"
+func (c *updatePodCache) dequeue() *updateNPMPod {
+	pod := c.cache[c.queue[0]]
```
Comment (vakalapa) [inline npm/pkg/dataplane/types.go:120]: check if 0th element is present before calling for it

**Review by vakalapa — CHANGES_REQUESTED** [review]

File: npm/pkg/dataplane/types.go (hunk 5 above)
Comment (huntergregory) [reply npm/pkg/dataplane/types.go:123]: discussed offline. caller must lock updatePodCache and check isEmpty() before calling dequeue()

**Review by vakalapa — APPROVED** [review]

**Pull request merged**